projects/storefrontlib/src/cms-components/cart/add-to-cart/add-to-cart.component.ts

```diff
@@ -45,30 +45,38 @@ export class AddToCartComponent implements OnInit, OnDestroy {
       this.cartEntry$ = this.cartService.getEntry(this.productCode);
       this.hasStock = true;
     } else {
-      this.subscription = this.currentProductService
-        .getProduct()
-        .pipe(filter(Boolean))
-        .subscribe(product => {
-          this.productCode = product.code;
-
-          if (
-            product.stock &&
-            product.stock.stockLevelStatus !== 'outOfStock' &&
-            product.stock.stockLevel > 0
-          ) {
-            this.maxQuantity = product.stock.stockLevel;
-            this.hasStock = true;
-          } else {
-            this.hasStock = false;
-          }
-
-          this.cartEntry$ = this.cartService.getEntry(this.productCode);
-
-          this.cd.markForCheck();
-        });
+      this.getProductDetails();
     }
   }
 
+  /**
+   * Uses the product code found in the URL params to retrieve the product details
+   */
+  getProductDetails(): void {
```

Comment:
I don't think this function should be public. It's an implementation detail rather than a good extension point. Convenience to use this function in unit tests isn't enough good reason to make it public.

To test reaction of the component to the changes of the stream `currentProductService.getProduct`, you may pass a custom `new BehaviorSubject(mockProduct)` into the spy of `currentProductService.getProduct` and later by push a new values into this stream (method `BehaviorSubject.next(mockProduct2)`)

Btw. please note that each execution of the function `getProductDetails` overwrites now a reference to a previous `subscription`, which leads into memory leaks. So I don't think we should have it as a separate function.

Reply:
Removed the function I created and changed the unit test. I did not use the BehaviorSubject as it would result in needing to change other tests as well. I think the implementation I went with is simple and should be good. Let me know what you think.

Also, thanks for informing me about the subscription leak. I'll be sure to look more into it and perhaps even speak to you about it at some point in the future to increase my knowledge.